Fix realign_raw and test_realign

qian-chu opened the pull request:
#### What does this implement/fix?
1. `realign_raw()` takes shared events in two raw objects and fits a polynomial to estimate offset and drift rate. However, when `other` is advanced in time than `raw` and needs to be cropped at the start, the function incorrectly uses `zero_ord` alone, while the actual offset for `other` should be `zero_ord / first_ord`.
2. When `other` contains channels that contain NaN (which is common in eye-tracking data, caused by blinks or losing track), resampling renders the entire channel data NaN. I added a warning with a reference to `mne.preprocessing.eyetracking.interpolate_blinks`.
3. The original code only cropped and resampled data but left the annotations unchanged. This works fine when events are embedded in stim channels as these got realigned, but in other cases (i.e., eye-tracking where annotations of saccades and fixations are stored in annotations), it would be convenient to also adjust annotations by multiplying onsets and durations to `first_ord`.

qian-chu commented:
Some additional tests:
Consider two objects where `raw` has a sampling rate of 1000Hz and several events that span four 1s long boxcars. `other` has a on-paper sampling rate of 1000Hz but actually 1100Hz, and some blank space at the start (1.1s) and end of the recording: (left is `raw`, right is `other`)
![image](https://github.com/mne-tools/mne-python/assets/97355086/396a6581-43d2-4baf-9105-1134be8cec8e)
Using `events_raw` and `events_other`, we can realign first using the old function. The `other` after realign_raw() looks like this:
![image](https://github.com/mne-tools/mne-python/assets/97355086/2140bbe2-13bb-4475-a0bd-ebd44480a92d)
Note that:
1. The onsets of the boxcars were not [1, 2, 3, 4] but ~90ms after. this is because the old cropping only cropped 1s which was the estimated `zero_ord`. Because of the drift, the time that needs to be cropped in `other` should be actually longer (1.1s, `zero_ord / first_ord`.
2. The events were unmodified, leading to significant deviation from the data and were thus not meaningful.

The realigned `other` after the proposed fix:
![image](https://github.com/mne-tools/mne-python/assets/97355086/64385e30-19f6-45e5-8486-38e1906a186a)
This should significantly facilitate realignment especially for eye-tracking data, where annotations contain useful information like saccade and fixation durations.

I'd love to incorporate this test into `test_realign.py` but currently I don't quite know how testing works. If someone can help me in this regard that would be great. I may also spend some time to learn to write proper tests when I have time.

Finally, thanks for any potential suggestions/reviews! This is my first contribution (to MNE and to any GitHub project) so please pardon me in advance if I'm doing anything wrong.

larsoner commented:
> Finally, thanks for any potential suggestions/reviews! This is my first contribution (to MNE and to any GitHub project) so please pardon me in advance if I'm doing anything wrong.

Sorry for the slow response, US holiday yesterday :slightly_smiling_face: 

Let me start off by saying this is awesome so far for a first contribution -- well described and justified fixes/changes! :+1: 

> I'd love to incorporate this test into test_realign.py but currently I don't quite know how testing works. If someone can help me in this regard that would be great. I may also spend some time to learn to write proper tests when I have time.

Feel free to read our [contributing guide](https://mne.tools/dev/install/contributing.html), but from a quick look I don't see where we talk about some basics, so I'll outline here:

1. We use `pytest`, which has lots of cool magic stuff that you'll probably figure out by seeing it in use2. If you are changing `mne.preprocessing.X` and the code lives in `mne/preprocessing/_X.py`, the tests you'll want to modify will probably live in `mne/preprocesseng/tests/test_X.py` or a similarly named file. So for you that's here:

   https://github.com/mne-tools/mne-python/blob/68ff5de17e4ca3cfe5a9ef9c8cf2c2473ec1d880/mne/preprocessing/tests/test_realign.py#L18

   You can see some cool magic in use in that function already in terms of `parametrize`.
3. Talking specifically about your idea -- you have to choose whether to make a new `def test_realign_annot` or modify the existing test function linked above. Given you have already written what appears to be a unit-test-like thing, I'd just add that as a new test, hopefully using just `RawArray` and `scipy.signal.square` etc. I think it would be clearer than modifying the existing test but I could be wrong!
4. To see if your test actually works, assuming you name it `test_realign_annot`, once you've written it you can run 

   ```
   $ pytest mne/preprocessing/tests/test_realign.py -k test_realign_annot
   ```
   And this will run just your new test. Once that works, you should run all realignment tests like `pytest mne/preprocessing/tests/test_realign.py`. In theory the code for that test should just be a trimmed-down version of whatever you used to make the nice square-wave plots above with the `plot` calls replaced by `assert_allclose` or `np.corrcoef` or whatever you need to do on the annotation data to say in numbers the same thing your plot tells our eyes.
4. It's possible that you'll need to tweak some values of some existing tests, hopefully reflecting *more correct* behavior in your PR.

Feel free to give it a shot and don't hesitate to ask for help if things aren't working!

qian-chu commented:
Thank you so much @larsoner for the detailed reply! Really appreciate your help especially regarding pytest. I will definitely look into it and improve the test. I've converted this PR into a draft until the test's fully functional.

qian-chu commented:
**Update**
I just updated `mne/preprocessing/tests/test_realign.py`. The original test simulated a stim channel and an EEG channel (hanning-smoothed impulse). I added another EEG channel with boxcar-like data and annotations that span each boxcar. Original objects (top=`raw`, bottom=`other`):
![image](https://github.com/mne-tools/mne-python/assets/97355086/920ed446-4663-48f8-aa2f-65621730f112)
![image](https://github.com/mne-tools/mne-python/assets/97355086/55d1fdf0-02b2-423d-b27d-a317a143338e)
After realignment (`other`):
![image](https://github.com/mne-tools/mne-python/assets/97355086/a51ce4eb-961c-45a6-b192-2e4d8992b4ed)

Higher drift rates (0.9 and 1.1) were added to pytest parameters to fully exhibit the advantages of the proposed fix. The old function should not be able to pass with such high drift rates.
Two tests were discarded:
https://github.com/mne-tools/mne-python/blob/9547c139c051f45440690c0b72da4c399b5a9368/mne/preprocessing/tests/test_realign.py#L64-L69
https://github.com/mne-tools/mne-python/blob/9547c139c051f45440690c0b72da4c399b5a9368/mne/preprocessing/tests/test_realign.py#L117-L119
They work when there is not much drift and thus are too restrictive. 

The key assertion for realign_annot is:
https://github.com/mne-tools/mne-python/blob/46f276e7c60d440f650598d45b1139f3e233edab/mne/preprocessing/tests/test_realign.py#L116-L135

All tests were successful on my computer:
![image](https://github.com/mne-tools/mne-python/assets/97355086/6ac0a891-138c-460d-9a88-84b6f07c5dd9)

drammock commented on mne/preprocessing/realign.py:
```diff
@@ -85,25 +91,41 @@ def realign_raw(raw, other, t_raw, t_other, verbose=None):
+    ]
+    if len(nan_ch_names) > 0:  # Issue warning if any channel in other has nan values
+        warn(
+            f"Channel(s) {', '.join(nan_ch_names)} in other contain NaN values\n"
+            + "Resampling these channels will result in empty data\n"
+            + "(If realigning eye-tracking data, "
+            + "consider using interpolate_blinks and passing interpolate_gaze=True)"
```
```suggestion
            f"Channel(s) {', '.join(nan_ch_names)} in `other` contain NaN values. "
            "Resampling these channels will result in the whole channel being NaN. "
            "(If realigning eye-tracking data, consider using interpolate_blinks and "
            "passing interpolate_gaze=True)"
```
You don't need `+` to concatenate strings across lines if they're within `(...)`.  Also we don't put newlines in most of our messages.

drammock commented on mne/preprocessing/realign.py:
```diff
@@ -85,25 +91,41 @@ def realign_raw(raw, other, t_raw, t_other, verbose=None):
 
-    # 4. Crop the end of one of the recordings if necessary
+    # 4. Realign the onsets and durations in other.annotations
+    # Must be ahead of end cropping to avoid losing annotations
```
```suggestion
    # Must happen before end cropping to avoid losing annotations
```

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -31,7 +35,10 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
+    sig_hann = np.convolve(signal, np.hanning(int(round(hann_len * sfreq))))[
+        : len(times)
+    ]
+    sig_box = np.convolve(signal, np.ones(int(round(box_len * sfreq))))[: len(times)]
```
what does this help test, that isn't already tested with `stim` (which is also a boxcar-like signal)?

qian-chu replied:
boxcars are meant to provide a data-wise "ground-truth" for the annotations. So that if we see the annotations line up with the boxcar signals we are sure we are doing fine

drammock replied:
ok, so the annotations are made to coincide with the boxcars, and you're testing that they remain co-incident.  So, can we instead eliminate `stim` from the test then?  Or is it not redundant?

qian-chu replied:
`stim` is for `find_event`, which subsequently defines `annotations`. I think it's nice to keep all three channels (`stim` to show event information is preserved after realignment; `hann` to simulate a naive EEG response; `boxcar` to coincide with annotations)

qian-chu replied:
I added an additional annotation/boxcar alignment test (at boxcar onsets boxcar data should transit from 0 to 1, and from 1 to 0 at offsets)

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -45,45 +52,87 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
+            for d, kind in (
+                (stim, "nearest"),
+                (sig_hann, "linear"),
+                (sig_box, "linear"),
```
boxcar interpolation should use `"nearest"` (if there's reason to keep `sig_box` in addition to `stim`)

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -45,45 +52,87 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
+            for d, kind in (
+                (stim, "nearest"),
+                (sig_hann, "linear"),
+                (sig_box, "linear"),
```
ditto

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -45,45 +52,87 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
 
-    # naive processing
     evoked_raw, events_raw, _, events_other = _assert_similarity(raw, other, n_events)
-    if start_raw == start_other:  # can just naively crop
-        a, b = data_raw[0], data_other[0]
-        n = min(len(a), len(b))
-        corr = np.corrcoef(a[:n], b[:n])[0, 1]
-        min_, max_ = (0.99999, 1.0) if sfreq_raw == sfreq_other else (0.8, 0.9)
-        assert min_ <= corr <= max_
```
I think we should try to keep this in if we can... you say this is discarded because it only works when there is not much drift.  Do the failures with larger drift still happen if you use "nearest" interpolation on `sig_box`?

qian-chu replied:
Yes the failure persists with nearest interpolation when drift is large. TBH I don't think the test makes much sense. It asserts that `raw` and `other` should contain somehow correlated data, which is clearly violated when there's drift. While `min_` and `_max` were adjusted accordingly in that case, the range was somewhat arbitrary (0.8 and 0.9)

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -45,45 +52,87 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
+            events_other[:, 0] / other.info["sfreq"],
+            atol=1 / sfreq,
+        )
+        onsets_raw, dur_raw = (
+            raw.annotations.onset - raw.first_samp / sfreq,
+            raw.annotations.duration,
+        )
+        onsets_other, dur_other = (
+            other.annotations.onset - other.first_samp / sfreq,
+            raw.annotations.duration,
+        )
```
This is hard to read, and one line per assignment would be shorter anyway. When reformatting it I noticed that the definition of `dur_other` is wrong
```suggestion
        onsets_raw = raw.annotations.onset - raw.first_samp / sfreq
        dur_raw = raw.annotations.duration
        onsets_other = other.annotations.onset - other.first_samp / sfreq
        dur_other = other.annotations.duration
```

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -113,7 +162,7 @@ def _assert_similarity(raw, other, n_events, events_raw=None):
-    corr = np.corrcoef(evoked_raw.data[0], evoked_other.data[0])[0, 1]
-    assert 0.9 <= corr <= 1.0
+    assert len(evoked_raw.data) == len(evoked_other.data) == 2  # just EEG
+    # corr = np.corrcoef(evoked_raw.data[0], evoked_other.data[0])[0, 1]
+    # assert 0.9 <= corr <= 1.0 # may not hold when drift is large
```
can we still test this but either use a lower threshold for large drifts, or at least do an `if`-clause so it still gets asserted for small drifts?

qian-chu replied:
Done. Although for the same reason as above, I don't think it's a meaningful test

drammock reviewed (commented):
Thanks for working on this!  Some comments/questions/suggestions below.

drammock commented on mne/preprocessing/realign.py:
```diff
@@ -31,6 +32,9 @@ def realign_raw(raw, other, t_raw, t_other, verbose=None):
         ``find_events(raw)[:, 0] - raw.first_event``.
     t_other : array-like, shape (n_events,)
         The times of shared events in ``other`` relative to ``other.times[0]``.
+    realign_annot : bool
+        If False, only apply realignment to data. If True, realign ``onset``
+        and ``duration`` in ``other.annotations`` as well.
```
@larsoner do you have an opinion here?  Is there any case where a user would want to realign the data and *not* realign its attached annotations?

larsoner replied:
I think what we do in main is wrong/a bug so no need for a new parameter

qian-chu replied:
Gotcha, I will eliminate the flag in the next commit

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -75,60 +74,104 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other, test
+    assert 0.98 < corr[0, 2] <= 1.0  # hanning
+    assert 0.98 < corr[1, 3] <= 1.0  # boxcar
+
+    # onsets drived from stim and annotations are the same
```
```suggestion
    # onsets derived from stim and annotations are the same
```

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -75,60 +74,104 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other, test
+    onsets_raw = raw.annotations.onset - first_time_raw
+    dur_raw = raw.annotations.duration
+
+    first_time_other = other.first_samp / other.info["sfreq"]
```
ditto

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -75,60 +74,104 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other, test
+
+    # onset/offset correspond to 0/1 transition in boxcar signals
+    n_events = len(onsets_raw)
+    onsets_samp_raw = (onsets_raw * sfreq).astype(int)
```
safer here to to `raw.time_as_index(onsets_raw)` to get the sample indices. That should also allow you to shrink the 10-sample window down to something much smaller.

qian-chu replied:
This is very good suggestion! It does shrink the sample window as well!

drammock commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -75,60 +74,104 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other, test
+    assert_allclose(
+        raw.get_data("raw_box")[0, onsets_samp_raw - 5],
+        [0] * n_events,
+        atol=0.2,  # atol to account for interpolation
```
shouldn't need any tolerance with `interpolation="nearest"`

qian-chu replied:
That interpolation comes from `other.load_data().resample` in `realign_raw`, not during the construction of `other`. If you look at the figures above there are some ripples caused by the filter

qian-chu commented:
Made some final modifications I can think of and I think it's in good shape for review/merge now! @larsoner @drammock I want to take a moment to thank you for your invaluable support and patient guidance during my first pull request. Your constructive feedback and timely responses made the experience both educational and encouraging for me. I truly appreciate it!

larsoner commented on mne/preprocessing/tests/test_realign.py:
```diff
@@ -45,45 +50,85 @@ def test_realign(ratio_other, start_raw, start_other, stop_raw, stop_other):
+    assert 0.99 < corr[0, 1] <= 1.0
+
+    # onsets derived from stim and annotations are the same
+    assert_allclose(
+        raw.annotations.onset, events_raw[:, 0] / raw.info["sfreq"], atol=2 / sfreq
+    )
```
this `atol` gets used a lot so it might be better to do
```suggestion
    atol = 2 / sfreq
    assert_allclose(
        raw.annotations.onset, events_raw[:, 0] / raw.info["sfreq"], atol=atol
    )
```
and use `atol=atol` below as well

larsoner commented on mne/preprocessing/realign.py:
```diff
@@ -85,25 +87,40 @@ def realign_raw(raw, other, t_raw, t_other, verbose=None):
-        logger.info(msg + "other")
-        other.crop(-zero_ord, None)
-        t_other += zero_ord
+        logger.info(f"Cropping {(zero_ord / first_ord):0.3f} s from the start of other")
+        other.crop(-(zero_ord / first_ord), None)
+        t_other += zero_ord / first_ord
```
The same computation is done three times (this requires effort to see and retype) so it seems a little cleaner to do the DRY version
```suggestion
        t_crop = zero_ord / first_ord
        logger.info(f"Cropping {t_crop:0.3f} s from the start of other")
        other.crop(-t_crop, None)
        t_other += t_crop
```

larsoner reviewed (approved):
Just some nitpicks but otherwise looks clean to me, @drammock feel free to merge if you're happy!

drammock commented:
looks good, marking for auto-merge (if the CIs all pass). Thanks @qian-chu!
